File: PackSocket/python/TcpPack.py
import struct
from socket import *
import threading
from threading import RLock
import time

REV_LEN = 1024
MAX_LISTEN_NUM = 20
GUID_LEN = 16
NewMsg_Size_LEN = 4
CMD_LEN = 4
Pos_cmd = 20
Pos_msg = 24
Pos_NewMsgMarkStrEnd = 1008
newMsgMaxDataSpace = 984
MarkLen = 1
ConnMark_Def = 0xFF


class newMsgMark:
    def __init__(self):
        super().__init__()
        self.NewMsgMarkStr = b"4A6959616E674C696E6DB62435625348"
        self.NewMsgMarkStrEnd = b"414B6D9D2AA1C14A6959616E674C696E"
    def isNewMsgMark(self,recvBuf):
        for i in range(GUID_LEN):
            if self.NewMsgMarkStr[i] != recvBuf[i]:
                 return False
        EndGDIDPos = REV_LEN - GUID_LEN
        for i in range(GUID_LEN):
            if self.NewMsgMarkStrEnd[i] != recvBuf[i + EndGDIDPos]:
                 return False
        return True
class Receiver:
    def __init__(self):
        self.__revSize = 0
        self.__msgSize = 0
        self.__cmd = 0
        self.__newMsgMark =  newMsgMark()
        self.__InitBuf()
    def __InitBuf(self,bufsize=REV_LEN):
        self.__buf = bytearray()
        for i in range(bufsize):
            self.__buf.append(0)

    def rev(self,recvBuf, RevFinishFun):
        if self.__newMsgMark.isNewMsgMark(recvBuf):
            self.__revSize = 0
            self.__msgSize = struct.unpack('<I', recvBuf[GUID_LEN:Pos_cmd])[0]
            self.__cmd = struct.unpack('<I', recvBuf[Pos_cmd:Pos_msg])[0]
            if len(self.__buf) < self.__msgSize: self.__InitBuf(self.__msgSize)
            if self.__msgSize <= newMsgMaxDataSpace:
                for i in range(0, self.__msgSize): self.__buf[i] = recvBuf[i+Pos_msg]
                self.__MsgFinish(RevFinishFun)
            return
        
        CpySize = REV_LEN
        if self.__revSize + REV_LEN > self.__msgSize:
            CpySize = self.__msgSize - self.__revSize

        for i in range(0,CpySize):self.__buf[self.__revSize+i] = recvBuf[i]
        self.__revSize = self.__revSize + CpySize
        
        if self.__revSize != self.__msgSize: return
        self.__MsgFinish(RevFinishFun)
        
    def __MsgFinish(self,RevFinishFun):
        RevFinishFun(self.__cmd,self.__buf, self.__msgSize)
        self.__msgSize = 0
        self.__revSize = 0

class Sender:
    def __InitBuf(self):
        self.__SendBuf = bytearray()
        for i in range(REV_LEN):
            self.__SendBuf.append(0)
    def __init__(self):
        self.__InitBuf()
        self.__newMsgMark = newMsgMark()

    def __Send(self,sock):
        by = bytes(self.__SendBuf)
        sock.send(by)

    def SendMsg(self,sock,cmd, msg):
        msgSize = len(msg)
        
        for i in range(0,GUID_LEN): self.__SendBuf[i] = self.__newMsgMark.NewMsgMarkStr[i]
        msgSizeBytes = struct.pack('<I',msgSize)
        for i in range(0,NewMsg_Size_LEN): self.__SendBuf[i+GUID_LEN] = msgSizeBytes[i]
        cmdBytes = struct.pack('<I',cmd)
        for i in range(0,CMD_LEN): self.__SendBuf[i+Pos_cmd] = cmdBytes[i]
        for i in range(0,GUID_LEN): self.__SendBuf[i + Pos_NewMsgMarkStrEnd] = self.__newMsgMark.NewMsgMarkStrEnd[i]
        if msgSize <= newMsgMaxDataSpace:
            for i in range(0,msgSize): self.__SendBuf[i+Pos_msg] = msg[i]
            self.__Send(sock)
            return
        self.__Send(sock)

        hasSendSize = 0
        while True:
            if msgSize <= REV_LEN:
                for i in range(0,msgSize): self.__SendBuf[i] = msg[i+hasSendSize]
                self.__Send(sock)
                return

            for i in range(0,REV_LEN): self.__SendBuf[i] = msg[i+hasSendSize]
            self.__Send(sock)

            hasSendSize = hasSendSize + REV_LEN
            msgSize = msgSize - REV_LEN
class MsgCtl:
    def __init__(self):
        self.__Receiver =  Receiver()
        self.__Sender =  Sender()
    def Rev(self,recvBuf, RevFinishFun):
        self.__Receiver.rev(recvBuf, RevFinishFun)
    def Send(self,sock,cmd, msg):
        self.__Sender.SendMsg(sock, cmd,msg)
class Connect:
    def __init__(self,_conn, _IServerRev, _IClientRev):
        self.__MsgCtl = MsgCtl()
        self.__ConnMark = ConnMark_Def
        self.__newMsgMark = newMsgMark()

        self.__IServerRev = _IServerRev
        self.__IClientRev = _IClientRev

        self.connectNormal = True
        self.__conn = _conn
        
        if None ==  self.__IServerRev and None ==  self.__IClientRev:
            return 
        self.__revThread = threading.Thread(target=self.__OnRev)  
        self.__revThread.setDaemon(True)  
        self.__revThread.start()
    def GetConnMark(self):
        return self.__ConnMark
        
    def Close(self):
        if(None != self.__conn): self.__conn.close() 
    def Rev(self, cmd, msg,  msgSize):
        if (None != self.__IServerRev): self.__IServerRev(self.__ConnMark, cmd,msg,msgSize)
        else: self.__IClientRev(cmd,msg,msgSize)
    def __OnRev(self):
        try:
            self.OnRev()
        except Exception as e:
            self.connectNormal = False
            self.Rev(0,None,0)
            print(self.__ConnMark)
            print(e)    

    def OnRev(self):
        recvBytes = bytearray()
        for i in range(REV_LEN):
            recvBytes.append(0)
        recBufPos = 0
        BufLen = 0

        if (self.__IServerRev != None):   
            markBytes =  self.__conn.recv(1)
            self.__ConnMark = markBytes[0]
            print("connect mark :" + str(self.__ConnMark))

        while(True):
            tempRevLen = REV_LEN - recBufPos
            tempRevBytes = self.__conn.recv(tempRevLen)            
            BufLen = len(tempRevBytes)
            for i in range(0,BufLen): recvBytes[recBufPos+i] = tempRevBytes[i]
            recBufPos += BufLen
            if recBufPos != REV_LEN: continue
            recBufPos = 0
            self.__MsgCtl.Rev(recvBytes,lambda cmd, revByte, revSize: self.Rev(cmd,revByte, revSize))
    
    def Send(self,cmd,msg):
        try:
            self.__MsgCtl.Send(self.__conn, cmd,msg)
        except Exception as e:
            print(e)
            return False
        return True


class SocketServer:
    def __init__(self):
        self.__SocketConnect = []
        self.__lock = RLock()
    def Start(self,port,  _revProc):
        self.__revProc = _revProc
        self.__socketServer = socket(AF_INET, SOCK_STREAM)
        self.__socketServer .bind(("0.0.0.0",port))
        
        threadSocket = threading.Thread(target=self.__Accept)  
        threadSocket.setDaemon(True)  
        threadSocket.start()
    def __Accept(self):
        self.__socketServer .listen(MAX_LISTEN_NUM)
        while True:
            clientsock,clientaddress=self.__socketServer.accept()
            print('connect from:',clientaddress)
            self.__lock.acquire()
            self.__lock.acquire()
            self.__RemoveErroConnect()
            conn = Connect(clientsock, self.__revProc, None)
            self.__SocketConnect.append(conn)
            self.__lock.release()
            self.__lock.release()
    def __RemoveErroConnect(self):
        self.__SocketConnect = list(filter(lambda x: x.connectNormal == True,self.__SocketConnect))

    def SendToAllConn(self,cmd,msg):
        self.__SendToConn(cmd,msg)   
    def SendToMarkConn(self, mark,cmd, msg):
        self.__SendToConn(cmd,msg, mark)
    def __SendToConn(self, cmd,msg,  mark = ConnMark_Def):
        self.__lock.acquire()
        self.__lock.acquire()
        for conn in self.__SocketConnect:
            if conn.connectNormal == False: continue

            if mark != ConnMark_Def:
                if mark != conn.GetConnMark(): continue
            
            conn.Send(cmd,msg)
            
        self.__lock.release()
        self.__lock.release()

class SocketClient:
    def __init__(self):
        self.__Connect = None
        self.__socketClient = None
        self.__MsgCtl = MsgCtl()
        self.__RevData = None

    def Start(self,mark,  ip,  port, IClientRev):
        self.__socketClient = socket(AF_INET, SOCK_STREAM)
        self.__socketClient.connect((ip,port)) 
        self.__socketClient.send(struct.pack('<B',mark))      
        self.__Connect = Connect(self.__socketClient, None, IClientRev)

    def Send(self,cmd,msg):
        if None == self.__Connect: return False
        return self.__Connect.Send(cmd,msg)


    def ProcRev(self):
        recvBytes = bytearray()
        for i in range(REV_LEN):
            recvBytes.append(0)
        recBufPos = 0
        BufLen = 0

        self.__RevData = None
        while(True):
            tempRevLen = REV_LEN - recBufPos
            tempRevBytes = self.__socketClient.recv(tempRevLen)            
            BufLen = len(tempRevBytes)
            for i in range(0,BufLen): recvBytes[recBufPos+i] = tempRevBytes[i]
            recBufPos += BufLen
            if recBufPos != REV_LEN: continue
            recBufPos = 0
            self.__MsgCtl.Rev(recvBytes,lambda cmd, revByte, revSize: self.__Rev(cmd,revByte, revSize))
            if None != self.__RevData:
                return self.__RevData
    def __Rev(self,cmd,revByte, revSize):
        self.__RevData = (cmd,revByte, revSize)

File: PackSocket/go/packsocket/tcppack.go
package packsocket

import (
	"bytes"
	"container/list"
	"encoding/binary"
	"fmt"
	"net"
	"sync"
	"time"
)

func uint32ToBytes(data uint32) []byte {
	bytebuf := bytes.NewBuffer([]byte{})
	binary.Write(bytebuf, binary.LittleEndian, data)
	return bytebuf.Bytes()
}
func bytesToUInt32(bys []byte, startpos int) int {
	bytebuff := bytes.NewBuffer(bys[startpos : startpos+4])
	var data uint32
	binary.Read(bytebuff, binary.LittleEndian, &data)
	return int(data)
}

const (
	socketRevBufSize = 1024

	guidLEN          = 16
	newMsgMarkStr    = "4A6959616E674C696E6DB62435625348"
	newMsgMarkStrEnd = "414B6D9D2AA1C14A6959616E674C696E"
	newMsgSizeLEN    = 4
	cmdLen           = 4

	posCmd                   = 20
	posMsg                   = 24
	posNewMsgMarkStrEnd      = 1008
	newMsgMaxDataSpace       = 984
	connMarkDef         byte = 255
)

//RevData revdata
type RevData struct {
	mark      byte
	cmd       uint32
	recvBytes []byte
	msglen    int
}

func (_this *RevData) init() {
	_this.mark = 0
	_this.cmd = 0
	_this.recvBytes = make([]byte, socketRevBufSize)
	_this.msglen = 0
}

//IRev imp
type IRev interface {
	onRec(data *RevData)
}

/////////receiver
type receiver struct {
	mRevSize       int
	mRevData       RevData
	mNewMsgMark    []byte
	mNewMsgMarkEnd []byte
}

func (_this *receiver) init() {
	_this.mRevSize = 0
	_this.mRevData.init()
	_this.mNewMsgMark = []byte(newMsgMarkStr)
	_this.mNewMsgMarkEnd = []byte(newMsgMarkStrEnd)
}
func (_this *receiver) rev(recvBuf []byte, Fun IRev) {
	if _this.isNewMsgMark(recvBuf) {
		_this.mRevSize = 0
		_this.mRevData.msglen = bytesToUInt32(recvBuf, guidLEN)
		_this.mRevData.cmd = uint32(bytesToUInt32(recvBuf, posCmd))
		if len(_this.mRevData.recvBytes) < int(_this.mRevData.msglen) {
			_this.mRevData.recvBytes = make([]byte, _this.mRevData.msglen)
		}

		if _this.mRevData.msglen <= newMsgMaxDataSpace {
			copy(_this.mRevData.recvBytes[0:_this.mRevData.msglen], recvBuf[posMsg:posMsg+_this.mRevData.msglen])
			_this.msgFinish(Fun)
		}
		return
	}

	CpySize := socketRevBufSize
	if _this.mRevSize+socketRevBufSize > _this.mRevData.msglen {
		CpySize = _this.mRevData.msglen - _this.mRevSize
	}

	copy(_this.mRevData.recvBytes[_this.mRevSize:], recvBuf[0:CpySize])
	_this.mRevSize = _this.mRevSize + CpySize

	if _this.mRevSize == _this.mRevData.msglen {
		_this.msgFinish(Fun)
	}
}
func (_this *receiver) isNewMsgMark(recvBuf []byte) bool {
	if !_this.checkGUID(recvBuf, _this.mNewMsgMark, 0) {
		return false
	}
	if !_this.checkGUID(recvBuf, _this.mNewMsgMarkEnd, posNewMsgMarkStrEnd) {
		return false
	}
	return true
}
func (_this *receiver) msgFinish(Fun IRev) {
	Fun.onRec(&_this.mRevData)
	_this.mRevData.msglen = 0
	_this.mRevSize = 0
}
func (_this *receiver) checkGUID(recvBuf []byte, guid []byte, posStart int) bool {
	for i := 0; i < guidLEN; i++ {
		if guid[i] != recvBuf[i+posStart] {
			return false
		}
	}
	return true
}

//////

///////////sender
type sender struct {
	mNewMsgMark    []byte
	mNewMsgMarkEnd []byte
}

func (_this *sender) init() {
	_this.mNewMsgMark = []byte(newMsgMarkStr)
	_this.mNewMsgMarkEnd = []byte(newMsgMarkStrEnd)
}
func (_this *sender) sendmsgFun(sock net.Conn, SendBuf []byte) bool {
	n, err := sock.Write(SendBuf)
	if err != nil {
		fmt.Println("sendmsg error:", err)
	}
	return socketRevBufSize == n
}
func (_this *sender) SendMsg(sock net.Conn, cmd uint32, msg []byte, msgSize int) bool {
	SendBuf := make([]byte, socketRevBufSize)

	copy(SendBuf[0:guidLEN], _this.mNewMsgMark[0:guidLEN])
	copy(SendBuf[guidLEN:guidLEN+newMsgSizeLEN], uint32ToBytes(uint32(msgSize))[0:newMsgSizeLEN])
	copy(SendBuf[posCmd:posCmd+cmdLen], uint32ToBytes(cmd)[0:cmdLen])
	copy(SendBuf[posNewMsgMarkStrEnd:posNewMsgMarkStrEnd+guidLEN], _this.mNewMsgMarkEnd[0:guidLEN])
	if msgSize <= newMsgMaxDataSpace {
		copy(SendBuf[posMsg:posMsg+msgSize], msg[0:msgSize])
		return _this.sendmsgFun(sock, SendBuf)
	}
	suc := _this.sendmsgFun(sock, SendBuf)
	if !suc {
		return false
	}

	hasSendSize := 0
	for {
		if msgSize <= socketRevBufSize {
			copy(SendBuf[0:msgSize], msg[hasSendSize:hasSendSize+msgSize])
			return _this.sendmsgFun(sock, SendBuf)
		}

		copy(SendBuf[0:socketRevBufSize], msg[hasSendSize:hasSendSize+socketRevBufSize])
		suc = _this.sendmsgFun(sock, SendBuf)
		if !suc {
			return false
		}

		hasSendSize += socketRevBufSize
		msgSize -= socketRevBufSize
	}
}

////////socketConn
type socketConn struct {
	mConn           net.Conn
	mPIServerRecver IRev
	mPIClientRecver IRev

	mSender   sender
	mReceiver receiver

	mBConnect bool

	mConnMark byte
}

func (_this *socketConn) onRec(data *RevData) {
	if nil != _this.mPIServerRecver {
		if nil != data {
			data.mark = _this.mConnMark
		}
		_this.mPIServerRecver.onRec(data)
	} else {
		_this.mPIClientRecver.onRec(data)
	}
}
func (_this *socketConn) revmsg(recvBuf []byte) int {
	revlen, err := _this.mConn.Read(recvBuf)
	if err != nil {
		fmt.Println("revmsg error:", err)
		return 0
	}
	return revlen
}
func (_this *socketConn) init(Conn net.Conn, pIServerRecver IRev, pIClientRecver IRev) {
	_this.mConn = Conn
	_this.mBConnect = true
	_this.mConnMark = connMarkDef
	_this.mPIServerRecver = pIServerRecver
	_this.mPIClientRecver = pIClientRecver
	_this.mSender.init()
	_this.mReceiver.init()
	if nil == pIServerRecver {
		if nil == pIClientRecver {
			return
		}
		go func() {
			recvBuf := make([]byte, socketRevBufSize)
			recBufPos := 0
			BufLen := 0

			if _this.mPIServerRecver != nil {
				BufLen = _this.revmsg(recvBuf[:1])
				if _this.recvErro(BufLen <= 0) {
					return
				}
				_this.mConnMark = recvBuf[0]
			}

			var Fun IRev = _this
			for {

				BufLen = _this.revmsg(recvBuf[recBufPos:])
				recBufPos += BufLen

				if _this.recvErro(BufLen <= 0) {
					return
				}

				if recBufPos != socketRevBufSize {
					continue
				}
				recBufPos = 0

				_this.mReceiver.rev(recvBuf, Fun)
			}
		}()
	}
}

func (_this *socketConn) recvErro(bErro bool) bool {
	if !bErro {
		return false
	}
	_this.mBConnect = false
	_this.onRec(nil)
	_this.mConn.Close()
	return true
}
func (_this *socketConn) Send(cmd uint32, msg []byte, msgSize int) bool {
	if !_this.mBConnect {
		return false
	}
	return _this.mSender.SendMsg(_this.mConn, cmd, msg, msgSize)
}

////////

//ServerSocket def
type ServerSocket struct {
	mSockSrv        net.Listener
	mPISocketRecver IRev

	mConnect *list.List
	mLock    *sync.Mutex
}

//Start socket
func (_this *ServerSocket) Start(port string, pISocketRecver IRev) bool {
	_this.mLock = new(sync.Mutex)

	_this.mConnect = list.New()

	_this.mPISocketRecver = pISocketRecver

	l, err := net.Listen("tcp", ":"+port)
	if err != nil {
		fmt.Println("listen error:", err)
		return false
	}

	go func() {
		for {
			conn, err := l.Accept()
			if err != nil {
				fmt.Println("accept error:", err)
				continue
			}

			_this.mLock.Lock()
			_this.removeInvalid()
			connect := new(socketConn)
			_this.mConnect.PushBack(connect)
			connect.init(conn, _this.mPISocketRecver, nil)
			_this.mLock.Unlock()
		}
	}()

	return true
}

//SendToAllConn all
func (_this *ServerSocket) SendToAllConn(cmd uint32, msg []byte, msgSize int) bool {
	return _this.sendToConn(cmd, msg, msgSize, connMarkDef)
}

//SendToMarkConn single
func (_this *ServerSocket) SendToMarkConn(mark byte, cmd uint32, msg []byte, msgSize int) bool {
	return _this.sendToConn(cmd, msg, msgSize, mark)
}
func (_this *ServerSocket) removeInvalid() {
	temp := list.New()
	for i := _this.mConnect.Front(); i != nil; i = i.Next() {
		var conn *socketConn = i.Value.(*socketConn)
		if conn.mBConnect {
			temp.PushBack(i)
		}
	}
	_this.mConnect = temp
}
func (_this *ServerSocket) sendToConn(cmd uint32, msg []byte, msgSize int, mark byte) bool {
	sendSuc := false
	_this.mLock.Lock()

	for i := _this.mConnect.Front(); i != nil; i = i.Next() {
		var conn *socketConn = i.Value.(*socketConn)
		if !conn.mBConnect {
			continue
		}

		if mark != connMarkDef {
			if mark != conn.mConnMark {
				continue
			}
		}

		sendSuc = conn.Send(cmd, msg, msgSize)
	}

	_this.mLock.Unlock()
	return sendSuc
}

//////

//ClientSocket def
type ClientSocket struct {
	mConnect  *socketConn
	mReceiver receiver
	mRevData  RevData
}

func (_this *ClientSocket) init() {
	_this.mConnect = nil
	_this.mReceiver.init()
}

//Start startClinet
func (_this *ClientSocket) Start(mark byte, IP string, port string, recver IRev) bool {
	conn, err := net.DialTimeout("tcp", IP+":"+port, 2*time.Second)
	if err != nil {
		fmt.Println("connect error:", err)
		return false
	}

	data := make([]byte, 1)
	data[0] = mark
	n, erro := conn.Write(data)
	if erro != nil {
		fmt.Println("connect error:", erro)
		return false
	}
	if n != 1 {
		return false
	}

	_this.mConnect = new(socketConn)
	_this.mConnect.init(conn, nil, recver)
	return true
}

//Send ClinetSendMsg
func (_this *ClientSocket) Send(cmd uint32, msg []byte, msgSize int) bool {
	if nil == _this.mConnect {
		return false
	}
	return _this.mConnect.Send(cmd, msg, msgSize)
}

//ProcRev callafterSend
func (_this *ClientSocket) ProcRev() *RevData {
	_this.mRevData.recvBytes = nil

	recvBuf := make([]byte, socketRevBufSize)
	recBufPos := 0
	BufLen := 0

	var Fun IRev = _this
	for {
		BufLen = _this.revmsg(recvBuf[recBufPos:])
		recBufPos += BufLen

		if BufLen <= 0 {
			return nil
		}

		if recBufPos != socketRevBufSize {
			continue
		}
		recBufPos = 0

		_this.mReceiver.rev(recvBuf, Fun)

		if _this.mRevData.recvBytes != nil {
			return &_this.mRevData
		}
	}
}
func (_this *ClientSocket) revmsg(recvBuf []byte) int {
	revlen, err := _this.mConnect.mConn.Read(recvBuf)
	if err != nil {
		fmt.Println("revmsg error:", err)
		return 0
	}
	return revlen
}

func (_this *ClientSocket) onRec(data *RevData) {
	_this.mRevData.recvBytes = data.recvBytes
	_this.mRevData.cmd = data.cmd
	_this.mRevData.msglen = data.msglen
}

////////
